core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2435,6 +2435,17 @@ private[zio] trait ZIOFunctions extends Serializable {
    */
   final val none: UIO[Option[Nothing]] = succeed(None)
 
+  /**
+   * Feeds elements of type A to f and collects all successes and failures in a tupled fashion.
+   */
+  final def partitionM[R, E, A, B](in: Iterable[A])(f: A => ZIO[R, E, B]): ZIO[R, Nothing, (List[E], List[B])] =
+    ZIO
+      .foreach(in)(f(_).either)
+      .map(_.foldRight((List.empty[E], List.empty[B])) {
```

Comment:
I like @adamgfraser 's solution, but I wonder if we can avoid its succinctness while avoiding the extra allocations:
```scala
ZIO.foldLeft(in)(List.empty[E] -> List.empty[B])((acc, x) =>
  f(x).fold(_ :: acc._1 -> acc._2, acc._1 -> _ :: acc._2))
).map(t => t._1.reverse -> t._2.reverse)
```
Granted, this doesn't yield a parallel version. Not even sure it would compile 😅 Just a thought.

Reply:
I can play around with this a bit more to try and figure out there is a solution more performant than Adams. I suppose we should always favour performant solutions over concise ones... @regiskuckaertz